### PR TITLE
M5 example

### PR DESCRIPTION
**Pull request by achoum**
- Create an example of processing for the M5 dataset.
- Add "todos" for when all the necessary operators are available.
- Register the drop_index and set_index operators
- Simplify the drop_index operator implementation. Fix issues with string features / indexes with multiple characters.
- Allow plotting with multiple events.
- Create a "unique_sampling" operator.
- Modify the "propagate" operator. Now, the new index is defined by a "sampling" argument (instead of features). This version is significantly simpler to use and efficient on the M5 dataset.

### CODE REVIEW

File: example/m5_competition.py [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+    tp.prefix("sum_28.", tp.moving_sum(sales, tp.duration.days(28))),
+    tp.prefix("sum_84.", tp.moving_sum(sales, tp.duration.days(84))),
+)
+
+# Lagged sales
+lagged_sales = []
+for lag in [1, 2, 3]:
+    lagged_sales.append(
+        tp.sample(
+            tp.prefix(f"lag_{lag}.", tp.lag(sales, tp.duration.days(lag))),
```
Comment (diegomarvid) [inline example/m5_competition.py:247]: Could we take advantage of the fact that `tp.lag()` can take as input an array of events?

Comment (achoum) [reply example/m5_competition.py:233]: From chat., I propose to keep it this way for now.

File: temporian/core/operators/drop_index.py [hunk 2]
```diff
@@ -137,35 +143,37 @@ def build_op_definition(cls) -> pb.OperatorDef:
...
         index_name
         for index_name in index_names
-        if index_name not in event.sampling.index
+        if index_name not in event.sampling.index.names
     ]
     if missing_index_names:
         raise KeyError(
-            f"{missing_index_names} are missing from the input event's index."
+            f"Dropped indexes {missing_index_names} are missing from the"
+            f" input index. The input index is {event.sampling.index}."
```
Comment (getchebarne) [inline temporian/core/operators/drop_index.py:168]: `{event.sampling.index} -> {event.sampling.index.names}` is more readable for this operator imo

File: temporian/core/operators/propagate.py [hunk 3]
```diff
@@ -104,55 +107,39 @@ def build_op_definition(cls) -> pb.OperatorDef:
...
+    """Propagates feature values over a larger index.
 
-    For example, suppose an index-less event sequence containing two features
-    "f_1" and "f_2", and containing 16 timestamps. Suppose "f_1" is a numerical
-    feature and "f_2" is a string feature with 4 unique values across the 16
-    timestamps. "Propagating" feature "f_1" over "f_2" will create an event
-    sequence containing feature "f_1" and indexed by feature "f_2". This event
-    sequence will contain 4 indexed time sequences each containing 16
-    timestamps.
+    Given "event" and "sampling" where "event" contains a super index of
```
Comment (getchebarne) [inline temporian/core/operators/propagate.py:114]: I think you meant `sub index` instead of `super index` here

File: temporian/implementation/numpy/data/plotter.py [hunk 4]
```diff
@@ -17,42 +19,52 @@ class Options(NamedTuple):
...
         height_per_plot_px: Height of each sub-plot (one per feature) in pixel.
         max_points: Maximum number of points to plot.
         min_time: If set, only plot events after min_time.
         max_time: If set, only plot events before min_time.
+        max_num_plots: Maximum number of plots to display. If more plots are
+          available, only plot the first "max_num_plots" ones and print a
+          warning.
     """
 
     if isinstance(event, list):
```
Comment (getchebarne) [inline temporian/implementation/numpy/data/plotter.py:52]: nit: rename `event` argument to `events`

File: temporian/implementation/numpy/operators/drop_index.py [hunk 5]
```diff
@@ -9,154 +9,123 @@
...
             if keep:
-                drop_feats = [
-                    NumpyFeature(
-                        idx_name,
-                        np.array(
-                            [src_index_lvl[i]] * this_block_length,
-                            dtype=DTYPE_REVERSE_MAPPING[dst_dtypes[idx_name]],
-                        ),
+                # Convert the dropped indexes into features
+                features = []
```
Comment (getchebarne) [inline temporian/implementation/numpy/operators/drop_index.py:68]: redundant w/ line 65 I think

File: temporian/implementation/numpy/operators/propagate.py [hunk 6]
```diff
@@ -18,63 +18,39 @@ def __init__(self, operator: Propagate) -> None:
...
-
-            num_values = len(src_to_item[0].data)
-            unique_index_values = set(
-                [
-                    tuple(
-                        [src_to_item[j].data[i] for j in range(num_new_index)]
-                    )
-                    for i in range(num_values)
-                ]
+        for sampling_index in sampling.sampling.data.keys():
```
Comment (getchebarne) [inline temporian/implementation/numpy/operators/propagate.py:26]: nit: this `.keys()` is unnecessary

File: temporian/implementation/numpy/operators/test/unique_timestamps_test.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+class UniqueTimestampsOperatorTest(absltest.TestCase):
+    def setUp(self):
+        pass
+
+    def test_base(self):
+        event_data = NumpyEvent.from_dataframe(
+            pd.DataFrame(
+                {
+                    "timestamp": [1, 2, 2, 2, 3, 3, 4],
```
Comment (getchebarne) [inline temporian/implementation/numpy/operators/test/unique_timestamps_test.py:34]: can we add a `3` to `c` index key `1` (so that we get multiple `3` timestamps, one for `c=1` and one for `c=2`)

**Review by getchebarne — APPROVED** [review]
really awesome work 👏 👏

File: benchmark/BUILD [hunk 8]
```diff
@@ -3,18 +3,23 @@ package(
...
 )
 
-sh_binary(
-    name = "time",
-    srcs = ["time.sh"],
-    data = [
-        "//benchmark/scripts",
-    ],
-)
+# TODO: Restore.
```
Comment (diegomarvid) [inline benchmark/BUILD:6]: reminder to restore this before merging

File: example/m5_competition.py [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+
+# Lagged sales
+lagged_sales = []
+for lag in [1, 2]:
+    lagged_sales.append(
+        tp.sample(
+            tp.prefix(f"lag_{lag}.", tp.lag(sales, tp.duration.days(lag))),
+            sales,
+        )
+    )
```
[the remark this answers is not in the document]
Comment (diegomarvid) [reply example/m5_competition.py:236]: looks good, would change the variable `lags` to something like `lags_in_days` just to express the unit of the lag and be more clear.

File: temporian/core/operators/glue.py [hunk 10]
```diff
@@ -23,7 +23,7 @@
 from temporian.proto import core_pb2 as pb
 
 # Maximum number of arguments taken by the glue operator
-MAX_NUM_ARGUMENTS = 10
+MAX_NUM_ARGUMENTS = 30
```
Comment (diegomarvid) [inline temporian/core/operators/glue.py:26]: very useful if we want to glue 28 lags. I had an issue for this #88 .

**Review by diegomarvid — APPROVED** [review]
Awesome! 🎉

File: example/m5_competition.py [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+    os.path.join(work_directory, "tabular_dataset.png")
+)
+
+# Export to csv file
+print("Export to csv file")
+print("==================")
+
+tabular_dataset_data.to_dataframe().to_csv(
+    os.path.join(work_directory, "tabular_dataset.csv"), index=False
+)
```
Comment (DonBraulio) [inline example/m5_competition.py:311]: nit: print a message saying that plots and csvs were saved to `work_directory` at the end.

File: example/m5_competition.py [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+sales_raw.info()
+print("sell_prices_raw:")
+sell_prices_raw.info()
+print("calendar_raw:")
+calendar_raw.info()
+
+# Melt dataset
+#
+# The different timesteps of sale data ("sales_raw") are stored in different
+# columns. Instead, we want for each timesteps to be a different row.
```
Comment (DonBraulio) [inline example/m5_competition.py:74]: Maybe adding this comment line also helps clarifying:
```
# 
Convert columns:
id,item_id,dept_id...d_1,d_2,d_3...d_n -> id,item_id,dept_id...day,sales
```

File: example/m5_competition.py [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+
+sales_raw = pd.melt(
+    sales_raw,
+    var_name="day",
+    value_name="sales",
+    id_vars=["id", "item_id", "dept_id", "cat_id", "store_id", "state_id"],
+)
+sales_raw["day"] = sales_raw["day"].apply(lambda x: int(x[2:]))
+
+print("sales_raw:\n", sales_raw)
```
Comment (DonBraulio) [inline example/m5_competition.py:87]: I'd use a different name here, like `sales_melt`, to find it easier at the output.

File: example/m5_competition.py [hunk 14]
```diff
@@ -0,0 +1,311 @@
...
+
+# Compute timestamps
+#
+# In the M5 raw dataset, timestamps are stored in three different ways in the
+# three different tables. To make our life easy, we will convert all the
+# timestamps into python datetimes objects.
+
+print("Compute timestamps")
+print("==================")
```
Comment (DonBraulio) [inline example/m5_competition.py:96]: I'd add to some or all of these lines line breaks: `print("\n\nCompute timestamps")`, the output has many lines and I found it a bit hard to parse.

File: example/m5_competition.py [hunk 15]
```diff
@@ -0,0 +1,311 @@
...
+print("Load raw dataset")
+print("================")
+
+# During development, set nrows to a small value (e.g. nrows=100) to
+# only load a sample of the dataset.
+#
+# TODO: Set to None when fast enough.
+nrows = 100
+
+raw_path = lambda x: os.path.join(raw_data_dir, x)
```
Comment (DonBraulio) [inline example/m5_competition.py:59]: Is there a reason to avoid `pathlib`? I guess we'd save some lines here and by doing just `savefig(work_directory / "raw_sales.png")` below.

File: example/m5_competition.py [hunk 16]
```diff
@@ -0,0 +1,311 @@
...
+lagged_sales = tp.glue(*lagged_sales)
+
+
+calendar_events = tp.glue(
+    tp.calendar_day_of_week(sales),
+    tp.calendar_day_of_month(sales),
+    tp.calendar_month(sales),
+)
+
+label_sales = []
```
Comment (DonBraulio) [inline example/m5_competition.py:246]: I'd change this name to `leak_sales`. If I'm missing something (I don't understand why it's `label_sales`), maybe add an explanatory comment?

**Review by DonBraulio — APPROVED** [review]
Nice one 💯 
I left minor comments, feel free to ignore them and we can prettify later.

File: example/m5_competition.py [hunk 17]
```diff
@@ -0,0 +1,311 @@
...
+"""
+
+import os
+import temporian as tp
+import pandas as pd
+import numpy as np
+from datetime import datetime, timezone, timedelta
+import urllib.request
+import zipfile
+import tempfile
```
[the remark this answers is not in the document]
Comment (achoum) [reply example/m5_competition.py:26]: *Done.
Maybe we can configure our formatter to do that automatically.

File: example/m5_competition.py [hunk 18]
```diff
@@ -0,0 +1,311 @@
...
+# dataset.
+work_directory = os.path.join(tempfile.gettempdir(), "m5")
+os.makedirs(work_directory, exist_ok=True)
+print("Use work directory:", work_directory)
+
+# Download the M5 dataset
+raw_data_zip = os.path.join(work_directory, "raw.zip")
+if not os.path.exists(raw_data_zip):
+    print("Download M5 dataset in", raw_data_zip)
+    url = "https://docs.google.com/uc?export=download&id=1NYHXmgrcXg50zR4CVWPPntHx9vvU5jbM&confirm=t"
```
[the remark this answers is not in the document]
Comment (achoum) [reply example/m5_competition.py:38]: No. This is the copy that someone from our group created for Kipu (the same used in the notebook).
*Adding comment

File: example/m5_competition.py [hunk 19]
```diff
@@ -0,0 +1,311 @@
...
+print("Melt dataset")
+print("============")
+
+sales_raw = pd.melt(
+    sales_raw,
+    var_name="day",
+    value_name="sales",
+    id_vars=["id", "item_id", "dept_id", "cat_id", "store_id", "state_id"],
+)
+sales_raw["day"] = sales_raw["day"].apply(lambda x: int(x[2:]))
```
[the remark this answers is not in the document]
Comment (achoum) [reply example/m5_competition.py:85]: I use this example for us to test and experiment with our API during development, I would keep the tutorial aspect for an actual tutorial (probably in a notebook with plenty of explanations and intermediate plotting).

Regarding the melting, I am not fully convinced we need to have it implemented in Temporian. Let's brainstorm about it on the next meeting.

File: example/m5_competition.py [hunk 20]
```diff
@@ -0,0 +1,311 @@
...
+wm_yr_wk_to_date = calendar_raw[calendar_raw["weekday"] == "Saturday"][
+    ["timestamp", "wm_yr_wk"]
+]
+map_wm_yr_wk_to_date = {}
+for _, row in wm_yr_wk_to_date.iterrows():
+    map_wm_yr_wk_to_date[row["wm_yr_wk"]] = row["timestamp"]
+
+sell_prices_raw["timestamp"] = sell_prices_raw["wm_yr_wk"].apply(
+    lambda x: map_wm_yr_wk_to_date[x]
+)
```
[the remark this answers is not in the document]
Comment (achoum) [reply example/m5_competition.py:124]: Ack

File: temporian/core/operators/propagate.py (hunk 3 above)
Comment (achoum) [reply temporian/core/operators/propagate.py:114]: In the context of set theory, and given we talk about the ensemble of index levels, I think the "super set" is correct here.
We still need to create a proper glossary with our terminology (and make sure the doc follows it) :)

File: temporian/implementation/numpy/data/plotter.py (hunk 4 above)
Comment (achoum) [reply temporian/implementation/numpy/data/plotter.py:52]: Done

File: temporian/implementation/numpy/operators/drop_index.py (hunk 5 above)
Comment (achoum) [reply temporian/implementation/numpy/operators/drop_index.py:68]: Good catch

File: temporian/implementation/numpy/operators/propagate.py (hunk 6 above)
Comment (achoum) [reply temporian/implementation/numpy/operators/propagate.py:26]: *Fixed

File: temporian/implementation/numpy/operators/test/unique_timestamps_test.py (hunk 7 above)
Comment (achoum) [reply temporian/implementation/numpy/operators/test/unique_timestamps_test.py:34]: *Done

File: temporian/core/operators/glue.py (hunk 10 above)
Comment (achoum) [reply temporian/core/operators/glue.py:26]: :D

I wonder if we need a limit though (I am not a fan to arbitrary limits like that).

File: example/m5_competition.py (hunk 11 above)
Comment (achoum) [reply example/m5_competition.py:311]: Ack

File: example/m5_competition.py (hunk 12 above)
Comment (achoum) [reply example/m5_competition.py:74]: *Done

File: example/m5_competition.py (hunk 13 above)
Comment (achoum) [reply example/m5_competition.py:87]: Let's figure the best naming for the tutorial.

File: example/m5_competition.py (hunk 14 above)
Comment (achoum) [reply example/m5_competition.py:96]: Meta-comment: The objective of this file is for us to test and improve our API. This is not a tutorial, and it will have to be re-worked before being something for the users. All the raised points are good, and I keep them for this re-work.

File: example/m5_competition.py (hunk 15 above)
Comment (achoum) [reply example/m5_competition.py:59]: Ack

File: example/m5_competition.py (hunk 16 above)
Comment (achoum) [reply example/m5_competition.py:246]: Ack

File: temporian/core/data/sampling.py [hunk 21]
```diff
@@ -80,29 +81,43 @@ def __iter__(self) -> Iterator[IndexLevel]:
     def levels(self) -> List[IndexLevel]:
         return self._levels
 
+    # TODO: Remove property
```
[the remark this answers is not in the document]
Comment (achoum) [reply temporian/core/data/sampling.py:84]: But then, the data is duplicated (more memory consumption, higher complexity to keep all the copies of the data in sync).

File: example/m5_competition.py [hunk 22]
```diff
@@ -0,0 +1,311 @@
...
+# Let's convert them into python datetimes.
+
+# TODO: Check if date conversion is correct, regarding UTC.
+calendar_raw["timestamp"] = calendar_raw["date"].apply(
+    lambda x: datetime.strptime(x, "%Y-%m-%d")
+)
+
+# In "sell_prices_raw", timestamps as expressed with a special index "wm_yr_wk"
+# defined in "calendar_raw". Let's convert them into python datetimes.
+wm_yr_wk_to_date = calendar_raw[calendar_raw["weekday"] == "Saturday"][
```
[the remark this answers is not in the document]
Comment (achoum) [reply example/m5_competition.py:126]: This is because the "wm_yr_wk" is updated once a week on Saturday.

File: example/m5_competition.py [hunk 23]
```diff
@@ -0,0 +1,311 @@
...
+del calendar_raw["wday"]
+del calendar_raw["month"]
+del calendar_raw["year"]
+
+# TODO: Use int16 when available.
+sales_raw["sales"] = sales_raw["sales"].astype(np.int32)
+# TODO: Use boolean when available.
+calendar_raw["snap_CA"] = calendar_raw["snap_CA"].astype(np.int32)
+calendar_raw["snap_TX"] = calendar_raw["snap_TX"].astype(np.int32)
+calendar_raw["snap_WI"] = calendar_raw["snap_WI"].astype(np.int32)
```
[the remark this answers is not in the document]
Comment (achoum) [reply example/m5_competition.py:163]: Yes, I hope so.
Note: This PR was written before the cast operator :)

File: example/m5_competition.py [hunk 24]
```diff
@@ -0,0 +1,311 @@
...
+            tp.prefix(f"leak_{lag}.", tp.leak(sales, tp.duration.days(lag))),
+            sales,
+        )
+    )
+label_sales = tp.glue(*label_sales)
+
+# Sum of the sales, sampled each day, per departement.
+sales_per_dept = tp.drop_index(sales, "item_id")
+sampling_once_per_day = tp.unique_timestamps(sales_per_dept["item_id"])
+sum_dayly_sales_per_dept = tp.prefix(
```
[the remark this answers is not in the document]
Comment (achoum) [reply example/m5_competition.py:270]: Nice catch :)

File: example/m5_competition.py [hunk 25]
```diff
@@ -0,0 +1,311 @@
...
+        sales_per_dept["sales"],
+        tp.duration.days(28),
+        sampling_once_per_day,
+    ),
+)
+
+# For each item, add the sum of departement sales for this specific item.
+sales_aggregated_item_level = tp.sample(
+    tp.propagate(sum_dayly_sales_per_dept, sales), sales
+)
```
[the remark this answers is not in the document]
Comment (achoum) [reply example/m5_competition.py:282]: Ack

File: example/m5_competition.py [hunk 26]
```diff
@@ -0,0 +1,311 @@
...
+    label_sales,
+    augmented_sales,
+    calendar_events,
+)
+
+# We run the computation
+tabular_dataset_data = tp.evaluate(
+    tabular_dataset,
+    {
+        sales: sales_data,
```
[the remark this answers is not in the document]
Comment (achoum) [reply example/m5_competition.py:302]: Yes. This is the kind of improvement we need to brainstorm about. This is related to our plan for eager mode.

File: temporian/core/data/duration.py [hunk 27]
```diff
@@ -83,6 +83,8 @@ def convert_date_to_duration(date: Timestamp) -> Duration:
         return convert_numpy_datetime64_to_duration(date)
     if isinstance(date, datetime.datetime):
         return convert_datetime_to_duration(date)
+    if isinstance(date, datetime.date):
+        return convert_datetime_date_to_duration(date)
```
[the remark this answers is not in the document]
Comment (achoum) [reply temporian/core/data/duration.py:87]: convert_date_to_duration is an already existing function in this file.

File: temporian/core/data/sampling.py [hunk 28]
```diff
@@ -80,29 +81,43 @@ def __iter__(self) -> Iterator[IndexLevel]:
...
     @property
     def names(self) -> List[str]:
         return [index_level.name for index_level in self._levels]
 
+    # TODO: Remove property
     @property
     def dtypes(self) -> List[IndexDType]:
         return [index_level.dtype for index_level in self._levels]
 
+    def __eq__(self, other):
```
[the remark this answers is not in the document]
Comment (achoum) [reply temporian/core/data/sampling.py:94]: *Done

File: temporian/core/data/sampling.py [hunk 29]
```diff
@@ -80,29 +81,43 @@ def __iter__(self) -> Iterator[IndexLevel]:
...
-            Index(index_levels)
-            if not isinstance(index_levels, Index)
-            else index_levels
-        )
+        if isinstance(index_levels, Index):
+            self._index = index_levels
+        else:
+            self._index = (
+                Index(index_levels)
+                if not isinstance(index_levels, Index)
```
[the remark this answers is not in the document]
Comment (achoum) [reply temporian/core/data/sampling.py:118]: Good point

File: temporian/core/operators/drop_index.py [hunk 30]
```diff
@@ -27,34 +27,30 @@ class DropIndexOperator(Operator):
...
-        # output sampling
+        output_features = self._output_features(event, index_to_drop, keep)
+
         output_sampling = Sampling(
             index_levels=[
-                (index_name, index_dtype)
-                for index_name, index_dtype in event.sampling.index
-                if index_name not in index_names
+                index_level
+                for index_level in event.sampling.index.levels
```
[the remark this answers is not in the document]
Comment (achoum) [reply temporian/core/operators/drop_index.py:49]: *Done
Then, we should remove the "levels" property. Having two way to do exactly the same operation is confusing :)

File: temporian/core/operators/drop_index.py [hunk 31]
```diff
@@ -65,54 +61,64 @@ def __init__(
...
-        ]
+        new_features = []
         if keep:
-            output_features = ([None] * len(index_names)) + output_features
-            for i, index_name in enumerate(index_names):
+            # Convert the index to drop into features.
+            #
+            # Note: The new features are added first.
+
+            new_features = []
```
[the remark this answers is not in the document]
Comment (achoum) [reply temporian/core/operators/drop_index.py:73]: Good catch

File: temporian/core/operators/drop_index.py (hunk 2 above)
Comment (achoum) [reply temporian/core/operators/drop_index.py:168]: *Done

**Comment by achoum** [comment]
Thanks
